src/quafu/tasks/task_database.py

```diff
@@ -0,0 +1,167 @@
+        if self.conn:
+            self.conn.commit()
+            self.conn.close()
+
+    def _create_table(self):
+        cursor = self.conn.cursor()
+        cursor.execute('''
+            CREATE TABLE IF NOT EXISTS tasks (
+                task_id TEXT PRIMARY KEY,
+                group_name TEXT DEFAULT NULL,
+                task_name TEXT DEFAULT NULL,
+                status TEXT,
+                priority INTEGER,
+                send_time TIMESTAMP,
+                finish_time TIMESTAMP DEFAULT NULL
+            )
+        ''')
+        cursor.close()
+
+    # region data manipulation
+    def insert_task(self,
+                    task_id,
+                    status,
+                    send_time: str,
+                    priority=2,
+                    group_name=None,
+                    task_name=None,
+                    finish_time: str = None
+                    ):
+        cursor = self.conn.cursor()
+        cursor.execute(
+            "INSERT INTO tasks "
+            "(task_id, group_name, task_name, status, priority, send_time, finish_time) "
+            "VALUES "
+            "(?, ?, ?, ?, ?, ?, ?)",
+            (task_id, group_name, task_name, status, priority, send_time, finish_time))
+        cursor.close()
+
+    def delete_task(self, task_id):
+        cursor = self.conn.cursor()
+        cursor.execute("DELETE FROM tasks WHERE task_id=?", (task_id,))
+        cursor.close()
+        print(f"Task {task_id} has been deleted from local database")
+
+    def update_task_status(self, task_id, status):
+        cursor = self.conn.cursor()
+        cursor.execute("UPDATE tasks SET status=? WHERE task_id=?", (status, task_id))
+        cursor.close()
+    # endregion
+
+    # region fetch tasks
+    def find_all_tasks(self):
+        cursor = self.conn.cursor()
+        cursor.execute("SELECT * FROM tasks")
+        tasks = cursor.fetchall()
+        cursor.close()
+        return tasks
+
+    def find_by_status(self, status):
+        cursor = self.conn.cursor()
+        cursor.execute("SELECT * FROM tasks WHERE status=?", (status,))
+        tasks = cursor.fetchall()
+        cursor.close()
+        return tasks
+
+    def find_by_priority(self, priority):
+        cursor = self.conn.cursor()
+        cursor.execute("SELECT * FROM tasks WHERE priority=?", (priority,))
+        tasks = cursor.fetchall()
+        cursor.close()
+        return tasks
+
+    def find_by_group(self, group_name):
+        cursor = self.conn.cursor()
+        if group_name is None:
+            cursor.execute("SELECT * FROM tasks WHERE group_name IS NULL")
+        else:
+            cursor.execute("SELECT * FROM tasks WHERE group_name=?", (group_name,))
+        tasks = cursor.fetchall()
+        cursor.close()
+        return tasks
+
+    def find_by_name(self, task_name):
+        cursor = self.conn.cursor()
+        if task_name is None:
+            cursor.execute("SELECT * FROM tasks WHERE task_name IS NULL")
```

Comment:
This use case seems strange. This means that when user wants to find tasks without taskname, they simpley call `find_by_name()` without any argument. IMHO, is it better to have a separate function like `find_tasks_no_name` or explicitly call `fina_by_name(task_name=None/"")`？

Moreover, if a user have a large number of tasks without task name (which I think could be a common case). If they simply call `find_by_name()`, they may be retrieving a large number of tasks from local database, will this harm the performance？Should we have a default setting of maximum number of tasks to retrieve?

Reply:
Just tested, there will be actually ``TypeError: QuafuTaskDatabase.find_by_name() missing 1 required positional argument: 'task_name'`` raised if there is no argument provided. So it has been required to explicitly call ``fina_by_name(task_name=None/"")``. (At least for ``python=3.11``?)

And I did not feel any time obstruct when extracting info of 576 tasks. Perhaps only when large data stored and exctracted,  we will need to consider efficiency. Let's discuss more detailly later.

Reply:
oh I see, thanks for the illustration